### PR TITLE
Implementing communication part for ihmc-ors-yarp library

### PR DESCRIPTION
**Pull request by gabrielenava**
I implemented the "send-robot-feedback" part. There are a couple of things I'm still trying to figure out, but I think the code as it is should work already. I 'm not sure I can test it right now (for that I think I need the `hello world` java program), but for debugging I printed one joint position after serialization, and it seems feedback is correctly serialized.

### CODE REVIEW

File: robots/icubGazeboSim/launch-bridgeihmcors-sim.xml [hunk 1]
```diff
@@ -50,8 +50,9 @@
 
     <!-- actual bridgeIHMCORS device -->
     <device name="ihmc_bridge" type="bridge_ihmc_ors">
-         <param name="period">0.005</param>
-
+         <param name="period">0.005</param> 
+	 <param name="remote-address">localhost</param>
```
Comment (traversaro) [inline robots/icubGazeboSim/launch-bridgeihmcors-sim.xml]: tabs --> space

File: src/BridgeIHMCORS.cpp [hunk 2]
```diff
@@ -30,13 +31,35 @@ BridgeIHMCORS::~BridgeIHMCORS()
...
+    yarp::os::Value address_val = config.check("remote-address",errAddr);    
+    
+    udp::resolver resolver(io_srv);
+    udp::resolver::query query(udp::v4(), address_val.asString(), portNumber.asString());  
+    udp::resolver::iterator iter = resolver.resolve(query);
+    udp::resolver::iterator end; 
+    
+    while (iter != end)
+    {
+    sender_endpoint = *iter++;
```
Comment (traversaro) [inline src/BridgeIHMCORS.cpp]: Please indent.

File: src/BridgeIHMCORS.cpp [hunk 3]
```diff
@@ -126,13 +148,15 @@ double rad2deg(const double angleInRad)
 
 void BridgeIHMCORS::run()
 {
+  
     yarp::os::LockGuard guard(m_deviceMutex);
-
+
```
Comment (traversaro) [inline src/BridgeIHMCORS.cpp]: If possible, try to avoid trailing whitespaces (most editor has an option to automatically remove them).

File: src/BridgeIHMCORS.h [hunk 4]
```diff
@@ -20,10 +20,20 @@
...
+// socket and port communication
+#include <iostream>
+#include <asio.hpp>
+
+// fast_cdr includes for serialization
+#include <fastcdr/FastBuffer.h>
+#include <fastcdr/Cdr.h>
+
+using namespace eprosima::fastcdr;
+using asio::ip::udp;
```
Comment (traversaro) [inline src/BridgeIHMCORS.h]: Not a big problem as this header will only be used in the BridgeIHMCORS.cpp compilation unit, but in general it is better to avoid using the `using` directives in headers (https://stackoverflow.com/questions/5849457/using-namespace-in-c-headers) .

File: src/BridgeIHMCORS.h [hunk 5]
```diff
@@ -34,6 +44,8 @@ namespace dev {
  * | Parameter name | SubParameter   | Type              | Units | Default Value | Required |   Description                                                     | Notes |
  * |:--------------:|:--------------:|:-----------------:|:-----:|:-------------:|:--------:|:-----------------------------------------------------------------:|:-----:|
  * | period         |      -         | double            |   s   | 0.005         | No       | Period at which the feedback collected by the robot devices is sent to the  IHMC-ORS controller | |
+ * | remote-address |      -         | string            |   -   |   -           | Yes      | IP address | |
+ * | remote-portNumber|    -         | string            |   -   |   -           | Yes      | Port number | |
```
Comment (traversaro) [inline src/BridgeIHMCORS.h]: Perhaps for consistency it is better `remote-port-number` ?

File: src/BridgeIHMCORS.cpp [hunk 6]
```diff
@@ -30,13 +31,35 @@ BridgeIHMCORS::~BridgeIHMCORS()
...
+    bool errPort = false;
+    
+    // TODO(nava): error message in case the parameter is not found
+    yarp::os::Value portNumber = config.check("remote-portNumber",errPort);
+    yarp::os::Value address_val = config.check("remote-address",errAddr);    
+    
+    udp::resolver resolver(io_srv);
+    udp::resolver::query query(udp::v4(), address_val.asString(), portNumber.asString());  
+    udp::resolver::iterator iter = resolver.resolve(query);
+    udp::resolver::iterator end;
```
Comment (traversaro) [inline src/BridgeIHMCORS.cpp]: It is not clear to me how this part works.

File: src/BridgeIHMCORS.cpp [hunk 7]
```diff
@@ -148,17 +172,37 @@ void BridgeIHMCORS::run()
...
-                // std::cerr << "m_robotFeedback message updated " << std::endl;
-            }
 
+           // serialize m_robotFeedback
+           Cdr m_robotFeedback_ser(fastBuffer);
+           m_robotFeedback_ser << m_robotFeedback;
+
+           // open socket
+           udp::socket socket(io_srv);
+           socket.open(udp::v4());
```
Comment (traversaro) [inline src/BridgeIHMCORS.cpp]: I am probably missing something: are you opening the socket at every loop?

**Review by traversaro — CHANGES_REQUESTED** [review]
I think something we did not consider is to think about is how to get the IHMC-ORS Java controller communicates to the bridge that it is up and that it is able to receive the streaming data. 
In general I would expect that the IHMC-ORS controller communicating with YARP requires the bridge to be up, not the other way around, but this will probably require some initial "configuration" communication between the controller and the bridge.

**Comment by traversaro** [comment]
By the way, you can also write a simple "mock" C++ program to test the communication, that pretends to be the IHMC controller, even without any need for using Java (if this simplifies the testing).

File: src/BridgeIHMCORS.cpp (hunk 7 above)
Comment (gabrielenava) [reply src/BridgeIHMCORS.cpp]: Actually, I tried to move these two lines in the header but I got an error. I will try again and debug.

File: src/BridgeIHMCORS.cpp [hunk 8]
```diff
@@ -147,16 +172,37 @@ void BridgeIHMCORS::run()
...
+               m_robotFeedback.jointStates()[jnt].tau() = m_jointTorquesFromYARP[jnt];
+               // std::cerr << "m_robotFeedback message updated " << std::endl;
+            } 
 
-                // std::cerr << "m_robotFeedback message updated " << std::endl;
-            }
 
+           // serialize m_robotFeedback
+           Cdr m_robotFeedback_ser(fastBuffer);
+           m_robotFeedback_ser << m_robotFeedback;
```
Comment (francesco-romano) [inline src/BridgeIHMCORS.cpp]: nitpick: use methods instead of operator if possible. The code will be clearer.
In this case, there is `serialize`

**Comment by gabrielenava** [comment]
I committed a  first version of the bridge with both the sender and receiver part.  It is possible to verify the data sent/received by the bridge by running `testSocket.cpp`, however the robot will not follow the desired torque values, because I didn't switch control mode to torque.

**Review by traversaro — APPROVED** [review]

**Comment by traversaro** [comment]
Merging to proceed with the integration, post-merging feedback is still welcome @francesco-romano @gabrielenava .

File: CMakeLists.txt [hunk 9]
```diff
@@ -14,7 +14,6 @@ find_package(fastcdr REQUIRED)
 
 # Find Asio and Threads, both required for opening sockets from C++
 find_package(Asio REQUIRED)
-find_package(Threads REQUIRED)
```
Comment (francesco-romano) [inline CMakeLists.txt]: Why did you remove this line?
I found that on some linux the `pthread` library is not automatically linked.

File: src/BridgeIHMCORS.h [hunk 10]
```diff
@@ -31,6 +32,38 @@
 namespace yarp {
 namespace dev {
 
+class BridgeIHMCORS;
+
+class BridgeIHMCORSReceiverThread : public yarp::os::Thread
+{
+    BridgeIHMCORS * m_parentDevice;
```
Comment (francesco-romano) [inline src/BridgeIHMCORS.h:39]: Given that the thread cannot exist without the parent device, I would make this explicit and use a reference.
Then, move the initialisation of the reference in the constructor instead of the configure

File: src/BridgeIHMCORS.cpp [hunk 11]
```diff
@@ -19,6 +19,77 @@ namespace yarp
...
+    m_desiredSocket->open(asio::ip::udp::v4());
+    m_desiredSocket->bind(m_desiredEndpoint);
+
+    return true;
+}
+
+void BridgeIHMCORSReceiverThread::resize(const int nrOfJoints)
+{
+    m_robotDesired.jointDesireds().resize(nrOfJoints);
+    m_desiredBuffer.resize(m_robotDesired.getMaxCdrSerializedSize());
```
Comment (francesco-romano) [inline src/BridgeIHMCORS.cpp:66]: `getMaxCdrSerializedSize` returns a very big number, independent from the number of joints.

File: src/BridgeIHMCORS.h (hunk 10 above)
Comment (traversaro) [reply src/BridgeIHMCORS.h:39]: I tend to avoid doing that because then you need to have a new/delete pair for allocating/deallocating the `BridgeIHMCORSReceiverThread`  in the device, but I don't see any big difference between the two approaches, so I can switch it to a reference.

File: src/BridgeIHMCORS.cpp (hunk 11 above)
Comment (traversaro) [reply src/BridgeIHMCORS.cpp:66]: Good catch, changing it to `getCdrSerializedSize` .
